FreeRTOS/FreeRTOS-Kernel
# Additional consistency checking in xPortSysTickHandler

Description
-----------
Background: We had a problem where we had a misconfigured SysTick priority coming from STCube (we are using an ST cortex-m microprocessor and their HAL libraries called STCube). For some reason they set the SysTick priority when setting up the clock three which must be done when coming back from stop mode. We had missed that the default value for this was 0 when porting code to a new processor. What happened was that we ended up in the forever for loop in list.c when inserting an item that is already there. The comment above the for loop then led us on a wild goose chase for almost three weeks. Which ended in us finally finding the bug and me and a college learned a lot of FreeRTOS internals (not a bad thing) :) 

Therefore the extra comment to help lead in the right direction when the SysTick is misconfigured. 

What we finally found was that when the error occurred the uxCriticalNesting was not zero when it should be. This was due to the fact that the SysTick interrupt had interrupted a critical section. Therefore the xPortSysTickHandler should check for this when configAssert is used.

I have added the configAssert in the ports where I feel confident it should be. Maybe it should be added to other ports as well. It would be good if someone with experience in those ports could take a look at that.

With these fixes our debugging time would have been hours instead of weeks.

Test Steps
-----------
To reproduce and test that it works simply set the SysTick priority to highest (logical) priority (0 numerical for cortex-m devices) and run the application. It will most likely crash very fast.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
